### PR TITLE
ci: port the appsec test ci to a dedicated github workflow

### PR DESCRIPTION
**Pull request by Julio-Guerra**
Introduce a new AppSec-specific test CI running on windows, macos, linux and docker containers.
It will run the combination of build tags and environment variables on the packages that matter to appsec.

The workflow only triggers on pull requests modifying appsec files, or with a manual run, or nightly.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -12,68 +12,68 @@ jobs:
...
-    - run:
-        name: milestone
-        command: |
-          go run checkmilestone.go
-
-    - run:
-        name: copyright
-        command: |
-          go run checkcopyright.go
+      - checkout
```
Comment (felixge) [inline .circleci/config.yml:15]: ❓ are these whitespace changes unrelated?

**Review by felixge — CHANGES_REQUESTED** [review]
I'm in support of this if it will be followed by more PRs to fully migrate us to Github Actions.

I'd be slightly against it it's going to stay as is.

Either way I think it'd be a good idea to make sure the results from this pipeline get uploaded to CI App. Otherwise this is a step backwards for the efforts by @maxday and myself to combat flaky tests.

File: .circleci/config.yml (hunk 1 above)
Comment (Julio-Guerra) [reply .circleci/config.yml]: The file isn't properly formatted yes. 
But I reverted them.

File: .github/actions/dd-ci-upload/action.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+# Composite action to upload junit test result files to Datadog CI App
```
Comment (felixge) [inline .github/actions/dd-ci-upload/action.yml:1]: 👏 This is cool. Didn't know you could include your own actions in like this!

**Review by felixge — APPROVED** [review]
LGTM based on the fact that this pipeline is only triggered by PRs using appsec files.

That being said, I hope we can migrate all our stuff to GH Actions in the next 3 months, and that @Hellzy will help in case this pipeline unexpectedly ends up causing issues for non-appsec related PRs.

File: .github/workflows/appsec.yml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  workflow_dispatch: # manually
+  schedule: # nightly
+    - cron: "0 0 * * *"
+  pull_request: # on pull requests touching appsec files
+    paths:
+      - 'internal/appsec/**'
+      - 'appsec/**'
+      - 'contrib/**/appsec.go'
+env:
+  DD_APPSEC_WAF_TIMEOUT: 5s
```
[the remark this answers is not in the document]
Comment (Hellzy) [reply .github/workflows/appsec.yml:12]: I'm curious about this as well. Are we expecting 1s of WAF execution to somehow be reached? 😮

**Review by Hellzy — APPROVED** [review]
LGTM overall.
I'm happy to take care of any issue that may arise because of this workflow if needed, as @felixge mentioned.

File: .github/workflows/appsec.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Julio-Guerra) [reply .github/workflows/appsec.yml:12]: No, it shouldn't be so long but we can't control the amount of CPU blocking time we get in those github runners. So 5s is chosen as randomly as 1s, the goal is just to never timeout, and not to validate it doesn't reach 1s.

File: .github/workflows/appsec.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  GODEBUG: cgocheck=2
+  JUNIT_REPORT: gotestsum-report.xml
+  TO_TEST: ./appsec/... ./internal/appsec/... ./contrib/google.golang.org/grpc/... ./contrib/net/http/... ./contrib/gorilla/mux/... ./contrib/go-chi/... ./contrib/labstack/echo.v4/... ./contrib/gin-gonic/gin/...
+jobs:
```
[the remark this answers is not in the document]
Comment (Julio-Guerra) [reply .github/workflows/appsec.yml:16]: Nice trace view, I didn't know that feature 👍 
So yes, from my experience, macOS and windows runners are super long to start.
As of job concurrency and so on, it depends on our actual github plan which I am unaware of. They are documented at
https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits

**Pull request merged**